<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en-us">
<head>
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  <meta property="og:title" content="The disaster that is Ruby&#39;s timeout method" />
<meta property="og:description" content="" />

<meta property="og:type" content="article" />

<meta property="og:locale" content="en_US" />
<meta property="og:url" content="http://localhost:1313/posts/rubys_timeout_method_broke_my_code/" />


  <title> The disaster that is Ruby&#39;s timeout method  &middot; vaneyckt.io </title>

  

  <link rel="stylesheet" href="/css/monokai.css">
  <script src="/js/highlight.pack.js"></script>

  <script>hljs.initHighlightingOnLoad();</script>

  
  <link rel="stylesheet" href="http://localhost:1313//css/poole.css">
  <link rel="stylesheet" href="http://localhost:1313//css/syntax.css">
  <link rel="stylesheet" href="http://localhost:1313//css/hyde.css">
  <link rel="stylesheet" href="http://fonts.googleapis.com/css?family=PT+Sans:400,400italic,700|Abril+Fatface">

  
  <link rel="apple-touch-icon-precomposed" sizes="144x144" href="/apple-touch-icon-144-precomposed.png">
  <link rel="shortcut icon" href="/favicon.ico">

  
  <link href="" rel="alternate" type="application/rss+xml" title="vaneyckt.io" />

  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css">

  <link href='http://fonts.googleapis.com/css?family=Raleway:400,300' rel='stylesheet' type='text/css'>

  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-71853042-1', 'auto');
    ga('send', 'pageview');
  </script>
</head>

<body>

<div class="sidebar">
  <div class="container sidebar-sticky">
    <div class="sidebar-about">
      <h1 class="brand"><a style="text-decoration:none" href="http://localhost:1313/">vaneyckt</a></h1>
      <p class="lead">
         notes to my future self 
      </p>
    </div>

    <ul class="sidebar-nav">
      <li><a href="http://localhost:1313/">Home</a></li>
      <li><a href="http://localhost:1313//posts">Posts</a></li>
      <li><a href="http://localhost:1313//topics">Tags</a></li>
      
      <br/>
      
    </ul>
      
      
      
      <a href="https://github.com/vaneyckt"><i class="fa fa-github-square"></i></a>&nbsp;&nbsp;
      <a href="mailto:tomvaneyck@gmail.com"><i class="fa fa-envelope-square"></i></a>&nbsp;&nbsp;
      <a href="http://vaneyckt.io/index.xml"><i class="fa fa-rss-square"></i></a>&nbsp;&nbsp;
      

    <p class="footnote">powered by <a href="http://hugo.spf13.com">Hugo</a> <br/>
    &copy; 2015 Tom Van Eyck. All rights reserved.</p>
  </div>
</div>


  <div class="content container">
    <div class="post">
      <h1 class="post-title">The disaster that is Ruby&#39;s timeout method</h1>
      <span class="post-date">Dec 10, 2015</span>
      

<p>On paper, <a href="http://ruby-doc.org/stdlib-2.1.1/libdoc/timeout/rdoc/Timeout.html">Ruby&rsquo;s timeout method</a> looks like an incredibly useful piece of code. Ever had a network request occasionally slow down your entire program because it just wouldn&rsquo;t finish? That&rsquo;s where <code>timeout</code> comes in. It provides a hard guarantee that a block of code will be finished within a specified amount of time.</p>

<pre><code class="language-ruby">require 'timeout'

status = Timeout::timeout(5) {
  # block of code that should be interrupted if it takes more than 5 seconds
}
</code></pre>

<p>There&rsquo;s one thing the documentation doesn&rsquo;t tell you though. If any of the lines in that block of code introduces side effects that rely on the execution of later lines of code to leave things in a stable state, then using the <code>timeout</code> method is a great way to introduce instability in your program. Examples of this include pretty much any program that is not entirely without stateful information. Let&rsquo;s have a closer look at this method to try and figure out what&rsquo;s going on here exactly.</p>

<h3 id="exceptions-absolutely-anywhere:6be3b8c16f7fbcee50b6dc9ac1c22c2e">Exceptions absolutely anywhere</h3>

<p>The problem with <code>timeout</code> is that it relies upon Ruby&rsquo;s questionable ability to have one thread raise an exception <em>absolutely anywhere</em> in an entirely different thread. The idea is that when you place code inside a <code>timeout</code> block, this code gets wrapped inside a new thread that executes in the background while the main thread goes to sleep for 5 seconds. Upon waking, the main thread grabs the background thread and forcefully stops it by raising an exception on it (<a href="https://github.com/ruby/ruby/blob/trunk/lib/timeout.rb#L72-L110">actual Ruby implementation</a>).</p>

<pre><code class="language-ruby"># raising_exceptions.rb
# threads can raise exceptions in other threads
thr = Thread.new do
  puts '...initializing resource'
  sleep 1

  puts '...using resource'
  sleep 1

  puts '...cleaning resource'
  sleep 1
end

sleep 1.5
thr.raise('raising an exception in the thread')
</code></pre>

<pre><code class="language-bash">$ ruby raising_exeptions.rb
$ ...initializing resource
$ ...using resource
</code></pre>

<p>The problem with this is that the main thread does not care what code the background thread is executing when it raises the exception. This means that the engineer responsible for the code that gets executed by the background thread needs to assume an exception can get thrown from <em>absolutely anywhere</em> within her code. This is madness! No one can be expected to place exception catchers around every single block of code!</p>

<p>The following code further illustrates the problem of being able to raise an exception <em>absolutely anywhere</em>. Turns out that <em>absolutely anywhere</em> includes locations like the inside of <code>ensure</code> blocks. These locations are generally not designed for handling any exceptions at all. I hope you weren&rsquo;t using an <code>ensure</code> block to terminate your database connection!</p>

<pre><code class="language-ruby"># ensure_block.rb
# raising exceptions inside an ensure block of another thread
thr = Thread.new do
  begin
    puts '...initializing resource'
    sleep 1

    raise 'something went wrong'

    puts '...using resource'
    sleep 1
  ensure
    puts '...started cleaning resource'
    sleep 1
    puts '...finished cleaning resource'
  end
end

sleep 1.5
thr.raise('raising an exception in the thread')

# prevent program from immediately terminating after raising exception
sleep 5
</code></pre>

<pre><code class="language-bash">$ ruby ensure_blocks.rb
$ ...initializing resource
$ ...started cleaning resource
</code></pre>

<h3 id="real-world-example:6be3b8c16f7fbcee50b6dc9ac1c22c2e">Real world example</h3>

<p>Recently, I spent a lot of time working with the <a href="https://github.com/taf2/curb">curb http client</a>. I ended up wrapping quite a few of my curb calls within <code>timeout</code> blocks because of tight time constraints. However, this caused great instability within the system I was working on. Sometimes a call would work, whereas other times that very same call would throw an exception about an invalid handle. It was this that caused me to start investigating the <code>timeout</code> method.</p>

<p>After having a bit of think, I came up with a proof of concept that showed beyond a doubt that the <code>timeout</code> method was introducing instability in the very internals of my http client. The finished proof of concept can feel a bit complex, so rather than showing the final code straightaway, I&rsquo;ll run you through my thought process instead.</p>

<p>Let&rsquo;s start with the basics and write some code that uses the http client to fetch a random google page. A randomized parameter is added to the google url in order to circumvent any client-side caching. The page fetch itself is wrapped inside a <code>timeout</code> block as we are interested in testing whether the <code>timeout</code> method is corrupting the http client.</p>

<pre><code class="language-ruby">require 'curb'
require 'timeout'

timeout(1) do
  Curl.get(&quot;http://www.google.com?foo=#{rand}&quot;)
end
</code></pre>

<p>This code will rarely timeout, as a page fetch generally takes way less than one second to complete. This is why we&rsquo;re going to wrap our page fetch inside a while loop.</p>

<pre><code class="language-ruby">require 'curb'
require 'timeout'

timeout(1) do
  while true
    Curl.get(&quot;http://www.google.com?foo=#{rand}&quot;)
  end
end
</code></pre>

<p>The above code is timing out and throwing a <code>Timeout::Error</code> exception. We now want to determine whether the timing out of one page fetch could corrupt the internal state of the http client, thereby causing problems for a subsequent page fetch. We&rsquo;ll need to make lots of page fetches in order to test this, so we&rsquo;re going to wrap all of our current code inside a while loop. Furthermore, we don&rsquo;t want any <code>Timeout::Error</code> exceptions to break us out of this while loop, so we&rsquo;re going to catch and ignore these exceptions inside the while loop we just created. This gives us our finished proof of concept code.</p>

<pre><code class="language-ruby"># proof_of_concept.rb
# timeout corrupts the very internals of the curb http client
require 'curb'
require 'timeout'

while true
  begin
    timeout(1) do
      while true
        Curl.get(&quot;http://www.google.com?foo=#{rand}&quot;)
      end
    end
  rescue Timeout::Error =&gt; e
  end
end
</code></pre>

<pre><code class="language-bash">$ ruby proof_of_concept.rb

/Users/vaneyckt/.rvm/gems/ruby-2.0.0-p594/gems/curb-0.8.8/lib/curl/easy.rb:67:
  in 'add': CURLError: The easy handle is already added to a multi handle
  (Curl::Err::MultiAddedAlready)
</code></pre>

<p>And there you have it. All of a sudden our http client is no longer reliable. This clearly illustrates that the <code>timeout</code> method caused an exception to be raised somewhere inside a critical code path of the http client, thereby corrupting the http client itself. It is this that caused the next page fetch to fail with the exception shown above.</p>

<h3 id="conclusion:6be3b8c16f7fbcee50b6dc9ac1c22c2e">Conclusion</h3>

<p>Hopefully this illustrates there is very little that you can do to prevent <code>timeout</code> from doing whatever it wants to your program&rsquo;s internal state. There is just no way a program can deal with exceptions being able to pop up <em>absolutely anywhere</em>. The only time you can truly get away with using timeouts is when you are writing purely functional code that does not rely on state. In all other cases it is best to just avoid using timeouts entirely.</p>

<p>Hopefully this illustrates that there is very little indeed that you can do to prevent <code>timeout</code> from doing whatever it wants to your program&rsquo;s internal state. There is just no way a program can deal with exceptions being able to pop up on literally any line. Just don&rsquo;t use timeout ever <nope. Well, it is possible in theory, but it would mean you would have to place exception catchers around every single block of code, as you don't know where the exception will be thrown. That's the problem here :(. It can be thrown from anywhere. http://jvns.ca/blog/2015/11/27/why-rubys-timeout-is-dangerous-and-thread-dot-raise-is-terrifying/ and https://www.reddit.com/r/programming/comments/3ui1sw/why_rubys_timeout_is_dangerous_and_threadraise_is/cxfg98b go into more detail than I'm doing here, but it is essentially impossible to deal with every possible edge case. .>. In fact, you should go and check your code for timeouts right now!</p>

    </div>
  </div>

<script data-no-instant>document.write('<script src="http://'
        + (location.host || 'localhost').split(':')[0]
		+ ':1313/livereload.js?mindelay=10"></'
        + 'script>')</script></body>
</html>
